File: app/src/main/java/com/ngengs/android/popularmovies/apps/data/local/MoviesProvider.java
package com.ngengs.android.popularmovies.apps.data.local;

import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.UriMatcher;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteQueryBuilder;
import android.net.Uri;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import android.util.Log;

import com.ngengs.android.popularmovies.apps.globals.LocalDatabase;
import com.ngengs.android.popularmovies.apps.utils.DatabaseHelper;

import java.util.Arrays;
import java.util.HashSet;

/**
 * Created by ngengs on 7/3/2017.
 */

@SuppressWarnings({"WeakerAccess", "unused"})
public class MoviesProvider extends ContentProvider {
    public static final String TAG = "MoviesProvider";


    private static final UriMatcher URI_MATCHER = buildUriMatcher();

    private static final String FAILED_TO_INSERT_ROW_INTO = "Failed to insert row into ";

    // movies._id = ?
    private static final String MOVIE_ID_SELECTION = MoviesEntry.TABLE_NAME + "." + MoviesEntry._ID + " = ? ";

    private DatabaseHelper databaseHelper;

    private static UriMatcher buildUriMatcher() {
        final UriMatcher uriMatcher = new UriMatcher(UriMatcher.NO_MATCH);
        final String authority = LocalDatabase.CONTENT_AUTHORITY;

        uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES, LocalDatabase.CODE_MOVIES);
        uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES + "/#", LocalDatabase.CODE_MOVIE_BY_ID);

        uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_POPULAR, LocalDatabase.CODE_POPULAR_MOVIES);
        uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_TOP_RATED, LocalDatabase.CODE_TOP_RATED_MOVIES);

        uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_FAVORITES, LocalDatabase.CODE_FAVORITES);

        return uriMatcher;
    }


    @Override
    public boolean onCreate() {
        databaseHelper = new DatabaseHelper(getContext());
        return false;
    }

    @Nullable
    @Override
    public Cursor query(@NonNull Uri uri, @Nullable String[] projection, @Nullable String selection, @Nullable String[] selectionArgs, @Nullable String sortOrder) {
        int match = URI_MATCHER.match(uri);
        Cursor cursor;
        checkColumns(match, projection);
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                cursor = databaseHelper.getReadableDatabase().query(
                        MoviesEntry.TABLE_NAME,
                        projection,
                        selection,
                        selectionArgs,
                        null,
                        null,
                        sortOrder
                );
                break;
            case LocalDatabase.CODE_MOVIE_BY_ID:
                cursor = getMovieById(uri, projection, sortOrder);
                break;
            case LocalDatabase.CODE_POPULAR_MOVIES:
                cursor = getMoviesFromReferenceTable(MoviesPopular.TABLE_NAME,
                        projection, selection, selectionArgs, sortOrder);
                break;
            case LocalDatabase.CODE_TOP_RATED_MOVIES:
                cursor = getMoviesFromReferenceTable(MoviesTopRated.TABLE_NAME,
                        projection, selection, selectionArgs, sortOrder);
                break;
            case LocalDatabase.CODE_FAVORITES:
                cursor = getMoviesFromReferenceTable(MoviesFavorites.TABLE_NAME,
                        projection, selection, selectionArgs, sortOrder);
                break;
            default:
                return null;
        }
        //noinspection ConstantConditions
        cursor.setNotificationUri(getContext().getContentResolver(), uri);
        return cursor;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        final int match = URI_MATCHER.match(uri);
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                return MoviesEntry.CONTENT_DIR_TYPE;
            case LocalDatabase.CODE_MOVIE_BY_ID:
                return MoviesEntry.CONTENT_ITEM_TYPE;
            case LocalDatabase.CODE_POPULAR_MOVIES:
                return MoviesPopular.CONTENT_DIR_TYPE;
            case LocalDatabase.CODE_TOP_RATED_MOVIES:
                return MoviesTopRated.CONTENT_DIR_TYPE;
            case LocalDatabase.CODE_FAVORITES:
                return MoviesFavorites.CONTENT_DIR_TYPE;
            default:
                return null;
        }
    }

    @Nullable
    @Override
    public Uri insert(@NonNull Uri uri, @Nullable ContentValues values) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        final int match = URI_MATCHER.match(uri);
        Uri returnUri;
        long id;
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                id = db.insertWithOnConflict(MoviesEntry.TABLE_NAME, null,
                        values, SQLiteDatabase.CONFLICT_REPLACE);
                if (id > 0) {
                    returnUri = MoviesEntry.buildMovieUri(id);
                } else {
                    throw new android.database.SQLException(FAILED_TO_INSERT_ROW_INTO + uri);
                }
                break;
            case LocalDatabase.CODE_POPULAR_MOVIES:
                id = db.insert(MoviesPopular.TABLE_NAME, null, values);
                if (id > 0) {
                    returnUri = MoviesPopular.CONTENT_URI;
                } else {
                    throw new android.database.SQLException(FAILED_TO_INSERT_ROW_INTO + uri);
                }
                break;
            case LocalDatabase.CODE_TOP_RATED_MOVIES:
                id = db.insert(MoviesTopRated.TABLE_NAME, null, values);
                if (id > 0) {
                    returnUri = MoviesTopRated.CONTENT_URI;
                } else {
                    throw new android.database.SQLException(FAILED_TO_INSERT_ROW_INTO + uri);
                }
                break;
            case LocalDatabase.CODE_FAVORITES:
                id = db.insert(MoviesFavorites.TABLE_NAME, null, values);
                if (id > 0) {
                    returnUri = MoviesFavorites.CONTENT_URI;
                } else {
                    throw new android.database.SQLException(FAILED_TO_INSERT_ROW_INTO + uri);
                }
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        //noinspection ConstantConditions
        getContext().getContentResolver().notifyChange(uri, null);
        return returnUri;
    }

    @Override
    public int delete(@NonNull Uri uri, @Nullable String selection, @Nullable String[] selectionArgs) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        final int match = URI_MATCHER.match(uri);
        int rowsDeleted;
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                rowsDeleted = db.delete(MoviesEntry.TABLE_NAME, selection, selectionArgs);
                break;
            case LocalDatabase.CODE_MOVIE_BY_ID:
                long id = MoviesEntry.getIdFromUri(uri);
                rowsDeleted = db.delete(MoviesEntry.TABLE_NAME, MOVIE_ID_SELECTION, new String[]{Long.toString(id)});
                break;
            case LocalDatabase.CODE_POPULAR_MOVIES:
                rowsDeleted = db.delete(MoviesPopular.TABLE_NAME, selection, selectionArgs);
                break;
            case LocalDatabase.CODE_TOP_RATED_MOVIES:
                rowsDeleted = db.delete(MoviesTopRated.TABLE_NAME, selection, selectionArgs);
                break;
            case LocalDatabase.CODE_FAVORITES:
                rowsDeleted = db.delete(MoviesFavorites.TABLE_NAME, selection, selectionArgs);
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        if (rowsDeleted != 0) {
            //noinspection ConstantConditions
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return rowsDeleted;
    }

    @Override
    public int update(@NonNull Uri uri, @Nullable ContentValues values, @Nullable String selection, @Nullable String[] selectionArgs) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        final int match = URI_MATCHER.match(uri);
        int rowsUpdated;
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                rowsUpdated = db.update(MoviesEntry.TABLE_NAME, values,
                        selection, selectionArgs);
                break;
            default:
                throw new UnsupportedOperationException("Unknown uri: " + uri);
        }
        if (rowsUpdated != 0) {
            //noinspection ConstantConditions
            getContext().getContentResolver().notifyChange(uri, null);
        }
        return rowsUpdated;
    }

    @Override
    public void shutdown() {
        databaseHelper.close();
        super.shutdown();
    }

    @Override
    public int bulkInsert(@NonNull Uri uri, @NonNull ContentValues[] values) {
        SQLiteDatabase db = databaseHelper.getWritableDatabase();
        final int match = URI_MATCHER.match(uri);
        int returnCount = 0;
        switch (match) {
            case LocalDatabase.CODE_MOVIES:
                db.beginTransaction();
                try {
                    for (ContentValues value : values) {
                        long id = db.insertWithOnConflict(MoviesEntry.TABLE_NAME,
                                null, value, SQLiteDatabase.CONFLICT_REPLACE);
                        if (id != -1) {
                            returnCount++;
                        }
                    }
                    db.setTransactionSuccessful();
                } finally {
                    db.endTransaction();
                }
                //noinspection ConstantConditions
                getContext().getContentResolver().notifyChange(uri, null);
                return returnCount;
            case LocalDatabase.CODE_POPULAR_MOVIES:
                db.beginTransaction();
                try {
                    db.setTransactionSuccessful();
                    for (ContentValues value : values) {
                        long id = db.insertWithOnConflict(MoviesPopular.TABLE_NAME, null, value, SQLiteDatabase.CONFLICT_REPLACE);
                        if (id != -1) {
                            returnCount++;
                        }
                    }
                } finally {
                    db.endTransaction();
                }
                //noinspection ConstantConditions
                getContext().getContentResolver().notifyChange(uri, null);
                return returnCount;
            case LocalDatabase.CODE_TOP_RATED_MOVIES:
                db.beginTransaction();
                try {
                    db.setTransactionSuccessful();
                    for (ContentValues value : values) {
                        long id = db.insertWithOnConflict(MoviesTopRated.TABLE_NAME, null, value, SQLiteDatabase.CONFLICT_REPLACE);
                        if (id != -1) {
                            returnCount++;
                        }
                    }
                } finally {
                    db.endTransaction();
                }
                //noinspection ConstantConditions
                getContext().getContentResolver().notifyChange(uri, null);
                return returnCount;
            default:
                return super.bulkInsert(uri, values);
        }
    }

    private Cursor getMovieById(Uri uri, String[] projection, String sortOrder) {
        long id = MoviesEntry.getIdFromUri(uri);
        return databaseHelper.getReadableDatabase().query(
                MoviesEntry.TABLE_NAME,
                projection,
                MOVIE_ID_SELECTION,
                new String[]{Long.toString(id)},
                null,
                null,
                sortOrder
        );
    }

    private Cursor getMoviesFromReferenceTable(String tableName, String[] projection, String selection,
                                               String[] selectionArgs, String sortOrder) {

        SQLiteQueryBuilder sqLiteQueryBuilder = new SQLiteQueryBuilder();

        // tableName INNER JOIN movies ON tableName.movie_id = movies._id
        sqLiteQueryBuilder.setTables(
                tableName + " INNER JOIN " + MoviesEntry.TABLE_NAME +
                        " ON " + tableName + "." + LocalDatabase.COLUMN_MOVIE_ID_KEY +
                        " = " + MoviesEntry.TABLE_NAME + "." + MoviesEntry._ID
        );
        Log.d(TAG, "getMoviesFromReferenceTable: " + sqLiteQueryBuilder.toString());

        return sqLiteQueryBuilder.query(databaseHelper.getReadableDatabase(),
                projection,
                selection,
                selectionArgs,
                null,
                null,
                sortOrder
        );
    }

    private void checkColumns(int match, String[] projection) {
        if (projection != null) {
            HashSet<String> availableColumns = new HashSet<>(Arrays.asList(MoviesEntry.getColumnsWithTable()));
            HashSet<String> requestedColumns = new HashSet<>(Arrays.asList(projection));
            if (!availableColumns.containsAll(requestedColumns)) {
                throw new IllegalArgumentException("Unknown columns in projection.");
            }
        }
    }
}

File: app/src/main/java/com/ngengs/android/popularmovies/apps/data/local/MoviesProvider.kt
package com.ngengs.android.popularmovies.apps.data.local

import android.content.ContentProvider
import android.content.ContentValues
import android.content.UriMatcher
import android.database.Cursor
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteQueryBuilder
import android.net.Uri
import android.provider.BaseColumns
import android.util.Log
import com.ngengs.android.popularmovies.apps.globals.LocalDatabase
import com.ngengs.android.popularmovies.apps.utils.DatabaseHelper
import kotlin.collections.HashSet

/**
 * Created by rizky.kharisma on 10/11/22.
 * @ngengs
 */
open class MoviesProvider : ContentProvider() {
    private lateinit var databaseHelper: DatabaseHelper

    override fun onCreate(): Boolean {
        databaseHelper = DatabaseHelper(context!!)
        return false
    }

    override fun query(
        uri: Uri,
        projection: Array<String?>?,
        selection: String?,
        selectionArgs: Array<String?>?,
        sortOrder: String?
    ): Cursor? {
        val match = URI_MATCHER.match(uri)
        checkColumns(match, projection)
        val cursor: Cursor? = when (match) {
            LocalDatabase.CODE_MOVIES -> databaseHelper.readableDatabase.query(
                MoviesEntry.TABLE_NAME,
                projection,
                selection,
                selectionArgs,
                null,
                null,
                sortOrder
            )
            LocalDatabase.CODE_MOVIE_BY_ID -> getMovieById(uri, projection, sortOrder)
            LocalDatabase.CODE_POPULAR_MOVIES -> getMoviesFromReferenceTable(
                MoviesPopular.TABLE_NAME,
                projection, selection, selectionArgs, sortOrder
            )
            LocalDatabase.CODE_TOP_RATED_MOVIES -> getMoviesFromReferenceTable(
                MoviesTopRated.TABLE_NAME,
                projection, selection, selectionArgs, sortOrder
            )
            LocalDatabase.CODE_FAVORITES -> getMoviesFromReferenceTable(
                MoviesFavorites.TABLE_NAME,
                projection, selection, selectionArgs, sortOrder
            )
            else -> return null
        }
        cursor?.setNotificationUri(context!!.contentResolver, uri)
        return cursor
    }

    override fun getType(uri: Uri): String? {
        return when (URI_MATCHER.match(uri)) {
            LocalDatabase.CODE_MOVIES -> MoviesEntry.CONTENT_DIR_TYPE
            LocalDatabase.CODE_MOVIE_BY_ID -> MoviesEntry.CONTENT_ITEM_TYPE
            LocalDatabase.CODE_POPULAR_MOVIES -> MoviesPopular.CONTENT_DIR_TYPE
            LocalDatabase.CODE_TOP_RATED_MOVIES -> MoviesTopRated.CONTENT_DIR_TYPE
            LocalDatabase.CODE_FAVORITES -> MoviesFavorites.CONTENT_DIR_TYPE
            else -> null
        }
    }

    override fun insert(uri: Uri, values: ContentValues?): Uri? {
        val db = databaseHelper.writableDatabase
        val match = URI_MATCHER.match(uri)
        val id: Long
        val returnUri: Uri = when (match) {
            LocalDatabase.CODE_MOVIES -> {
                id = db.insertWithOnConflict(
                    MoviesEntry.TABLE_NAME, null,
                    values, SQLiteDatabase.CONFLICT_REPLACE
                )
                if (id > 0) {
                    MoviesEntry.buildMovieUri(id)
                } else {
                    throw SQLException(FAILED_TO_INSERT_ROW_INTO + uri)
                }
            }
            LocalDatabase.CODE_POPULAR_MOVIES -> {
                id = db.insert(MoviesPopular.TABLE_NAME, null, values)
                if (id > 0) {
                    MoviesPopular.CONTENT_URI
                } else {
                    throw SQLException(FAILED_TO_INSERT_ROW_INTO + uri)
                }
            }
            LocalDatabase.CODE_TOP_RATED_MOVIES -> {
                id = db.insert(MoviesTopRated.TABLE_NAME, null, values)
                if (id > 0) {
                    MoviesTopRated.CONTENT_URI
                } else {
                    throw SQLException(FAILED_TO_INSERT_ROW_INTO + uri)
                }
            }
            LocalDatabase.CODE_FAVORITES -> {
                id = db.insert(MoviesFavorites.TABLE_NAME, null, values)
                if (id > 0) {
                    MoviesFavorites.CONTENT_URI
                } else {
                    throw SQLException(FAILED_TO_INSERT_ROW_INTO + uri)
                }
            }
            else -> throw UnsupportedOperationException("Unknown uri: $uri")
        }
        context!!.contentResolver.notifyChange(uri, null)
        return returnUri
    }

    override fun delete(uri: Uri, selection: String?, selectionArgs: Array<out String>?): Int {
        val db = databaseHelper.writableDatabase
        val match = URI_MATCHER.match(uri)
        val rowsDeleted: Int = when (match) {
            LocalDatabase.CODE_MOVIES -> db.delete(MoviesEntry.TABLE_NAME, selection, selectionArgs)
            LocalDatabase.CODE_MOVIE_BY_ID -> {
                val id = MoviesEntry.getIdFromUri(uri)
                db.delete(
                    MoviesEntry.TABLE_NAME,
                    MOVIE_ID_SELECTION,
                    arrayOf(java.lang.Long.toString(id))
                )
            }
            LocalDatabase.CODE_POPULAR_MOVIES -> db.delete(
                MoviesPopular.TABLE_NAME,
                selection,
                selectionArgs
            )
            LocalDatabase.CODE_TOP_RATED_MOVIES -> db.delete(
                MoviesTopRated.TABLE_NAME,
                selection,
                selectionArgs
            )
            LocalDatabase.CODE_FAVORITES -> db.delete(
                MoviesFavorites.TABLE_NAME,
                selection,
                selectionArgs
            )
            else -> throw java.lang.UnsupportedOperationException("Unknown uri: $uri")
        }
        if (rowsDeleted != 0) {
            context!!.contentResolver.notifyChange(uri, null)
        }
        return rowsDeleted
    }

    override fun update(
        uri: Uri,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<out String>?
    ): Int {
        val db = databaseHelper.writableDatabase
        val match = URI_MATCHER.match(uri)
        val rowsUpdated: Int = when (match) {
            LocalDatabase.CODE_MOVIES ->
                db.update(MoviesEntry.TABLE_NAME, values, selection, selectionArgs)
            else -> throw java.lang.UnsupportedOperationException("Unknown uri: $uri")
        }
        if (rowsUpdated != 0) {
            context!!.contentResolver.notifyChange(uri, null)
        }
        return rowsUpdated
    }

    override fun shutdown() {
        databaseHelper.close()
        super.shutdown()
    }


    override fun bulkInsert(uri: Uri, values: Array<ContentValues?>): Int {
        val db = databaseHelper.writableDatabase
        val match = URI_MATCHER.match(uri)
        var returnCount = 0
        return when (match) {
            LocalDatabase.CODE_MOVIES -> {
                db.beginTransaction()
                try {
                    for (value in values) {
                        val id = db.insertWithOnConflict(
                            MoviesEntry.TABLE_NAME,
                            null,
                            value,
                            SQLiteDatabase.CONFLICT_REPLACE
                        )
                        if (id != -1L) {
                            returnCount++
                        }
                    }
                    db.setTransactionSuccessful()
                } finally {
                    db.endTransaction()
                }
                context!!.contentResolver.notifyChange(uri, null)
                returnCount
            }
            LocalDatabase.CODE_POPULAR_MOVIES -> {
                db.beginTransaction()
                try {
                    db.setTransactionSuccessful()
                    for (value in values) {
                        val id = db.insertWithOnConflict(
                            MoviesPopular.TABLE_NAME,
                            null,
                            value,
                            SQLiteDatabase.CONFLICT_REPLACE
                        )
                        if (id != -1L) {
                            returnCount++
                        }
                    }
                } finally {
                    db.endTransaction()
                }
                context!!.contentResolver.notifyChange(uri, null)
                returnCount
            }
            LocalDatabase.CODE_TOP_RATED_MOVIES -> {
                db.beginTransaction()
                try {
                    db.setTransactionSuccessful()
                    for (value in values) {
                        val id = db.insertWithOnConflict(
                            MoviesTopRated.TABLE_NAME,
                            null,
                            value,
                            SQLiteDatabase.CONFLICT_REPLACE
                        )
                        if (id != -1L) {
                            returnCount++
                        }
                    }
                } finally {
                    db.endTransaction()
                }
                context!!.contentResolver.notifyChange(uri, null)
                returnCount
            }
            else -> super.bulkInsert(uri, values)
        }
    }

    private fun getMovieById(uri: Uri, projection: Array<String?>?, sortOrder: String?): Cursor? {
        val id = MoviesEntry.getIdFromUri(uri)
        return databaseHelper.readableDatabase.query(
            MoviesEntry.TABLE_NAME,
            projection,
            MOVIE_ID_SELECTION, arrayOf(id.toString()),
            null,
            null,
            sortOrder
        )
    }


    private fun getMoviesFromReferenceTable(
        tableName: String, projection: Array<String?>?, selection: String?,
        selectionArgs: Array<String?>?, sortOrder: String?
    ): Cursor? {
        val sqLiteQueryBuilder = SQLiteQueryBuilder()

        // tableName INNER JOIN movies ON tableName.movie_id = movies._id
        sqLiteQueryBuilder.tables = tableName + " INNER JOIN " + MoviesEntry.TABLE_NAME +
                " ON " + tableName + "." + LocalDatabase.COLUMN_MOVIE_ID_KEY +
                " = " + MoviesEntry.TABLE_NAME + "." + BaseColumns._ID
        Log.d(TAG, "getMoviesFromReferenceTable: $sqLiteQueryBuilder")
        return sqLiteQueryBuilder.query(
            databaseHelper.readableDatabase,
            projection,
            selection,
            selectionArgs,
            null,
            null,
            sortOrder
        )
    }

    private fun checkColumns(match: Int, projection: Array<String?>?) {
        if (projection != null) {
            val availableColumns = HashSet(listOf(MoviesEntry.getColumnsWithTable()))
            val requestedColumns = HashSet(listOf(projection.mapNotNull { it }.toTypedArray()))
//            require(availableColumns.containsAll(requestedColumns)) { "Unknown columns in projection." }
        }
    }

    companion object {
        const val TAG = "MoviesProvider"


        private val URI_MATCHER = buildUriMatcher()

        private const val FAILED_TO_INSERT_ROW_INTO = "Failed to insert row into "

        // movies._id = ?
        private const val MOVIE_ID_SELECTION =
            MoviesEntry.TABLE_NAME + "." + BaseColumns._ID + " = ? "


        private fun buildUriMatcher(): UriMatcher {
            val uriMatcher = UriMatcher(UriMatcher.NO_MATCH)
            val authority = LocalDatabase.CONTENT_AUTHORITY
            uriMatcher.addURI(authority, LocalDatabase.PATH_MOVIES, LocalDatabase.CODE_MOVIES)
            uriMatcher.addURI(
                authority,
                LocalDatabase.PATH_MOVIES + "/#",
                LocalDatabase.CODE_MOVIE_BY_ID
            )
            uriMatcher.addURI(
                authority,
                LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_POPULAR,
                LocalDatabase.CODE_POPULAR_MOVIES
            )
            uriMatcher.addURI(
                authority,
                LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_TOP_RATED,
                LocalDatabase.CODE_TOP_RATED_MOVIES
            )
            uriMatcher.addURI(
                authority,
                LocalDatabase.PATH_MOVIES + "/" + LocalDatabase.PATH_FAVORITES,
                LocalDatabase.CODE_FAVORITES
            )
            return uriMatcher
        }
    }
}